crypto/crmf/crmf_lib.c

```diff
@@ -0,0 +1,767 @@
+    }
+    return 0;
+}
+
+/* TODO: support cases 1+2 (besides case 3) defined in RFC 4211, section 4.1. */
+static int CRMF_poposigningkey_init(OSSL_CRMF_POPOSIGNINGKEY *ps,
+                                    OSSL_CRMF_CERTREQUEST *cr,
+                                    const EVP_PKEY *pkey, int dgst)
+{
+    int len;
+    size_t crlen, max_sig_size;
+    unsigned int siglen;
+    unsigned char *crder = NULL, *sig = NULL;
+    int alg_nid=0, md_nid=0;
+    const EVP_MD *alg = NULL;
+
+    EVP_MD_CTX *ctx = NULL;
+
+    if (ps == NULL || cr == NULL || pkey == NULL) {
+        CRMFerr(CRMF_F_CRMF_POPOSIGNINGKEY_INIT, CRMF_R_NULL_ARGUMENT);
+        return 0;
+    }
+
+    /* OpenSSL defaults all bit strings to be encoded as ASN.1 NamedBitList */
+    ps->signature->flags &= ~(ASN1_STRING_FLAG_BITS_LEFT | 0x07);
+    ps->signature->flags |= ASN1_STRING_FLAG_BITS_LEFT;
+
+    len = i2d_OSSL_CRMF_CERTREQUEST(cr, &crder);
+    if (len < 0 || crder == NULL)
+        goto err;
+    crlen = (size_t)len;
+
+    max_sig_size = EVP_PKEY_size((EVP_PKEY *)pkey);
+    sig = OPENSSL_malloc(max_sig_size);
+    if (sig == NULL)
+        goto err;
+
+    if (!OBJ_find_sigid_by_algs(&alg_nid, dgst, EVP_PKEY_id(pkey))) {
+        CRMFerr(CRMF_F_CRMF_POPOSIGNINGKEY_INIT,
+                CRMF_R_UNSUPPORTED_ALG_FOR_POPSIGNINGKEY);
+        goto err;
+    }
+    if (!(OBJ_find_sigid_algs(alg_nid, &md_nid, NULL)
+          && (alg = EVP_get_digestbynid(md_nid)) != NULL)) {
+        CRMFerr(CRMF_F_CRMF_POPOSIGNINGKEY_INIT,
+                CRMF_R_UNSUPPORTED_ALG_FOR_POPSIGNINGKEY);
+        goto err;
+    }
+    if (!X509_ALGOR_set0(ps->algorithmIdentifier, OBJ_nid2obj(alg_nid),
+                         V_ASN1_NULL, NULL)
+        || (ctx = EVP_MD_CTX_new()) == NULL
+        || !(EVP_SignInit_ex(ctx, alg, NULL)))
+        goto err;
+    if (!(EVP_SignUpdate(ctx, crder, crlen)))
+        goto err;
+    if (!(EVP_SignFinal(ctx, sig, &siglen, (EVP_PKEY *)pkey)))
+        goto err;
+
+    if (!(ASN1_BIT_STRING_set(ps->signature, sig, siglen)))
+        goto err;
+
+    /* cleanup */
+    OPENSSL_free(crder);
+    EVP_MD_CTX_free(ctx);
+    OPENSSL_free(sig);
+    return 1;
+ err:
+    CRMFerr(CRMF_F_CRMF_POPOSIGNINGKEY_INIT, CRMF_R_ERROR);
+    OPENSSL_free(crder);
+    if (ctx)
+        EVP_MD_CTX_free(ctx);
+    OPENSSL_free(sig);
+    return 0;
+}
+
+
+int OSSL_CRMF_MSG_create_popo(OSSL_CRMF_MSG *crm, const EVP_PKEY *pkey,
+                              int dgst, int ppmtd)
+{
+    OSSL_CRMF_POPO *pp = NULL;
+
+    if (crm == NULL
+        || (ppmtd == OSSL_CRMF_POPO_SIGNATURE && pkey == NULL)) {
+        CRMFerr(CRMF_F_OSSL_CRMF_MSG_CREATE_POPO, CRMF_R_NULL_ARGUMENT);
+        return 0;
+    }
+
+    if ((pp = OSSL_CRMF_POPO_new()) == NULL)
+        goto oom;
+    pp->type = ppmtd;
+
+    switch (ppmtd) {
+    case OSSL_CRMF_POPO_NONE:
+        OSSL_CRMF_POPO_free(pp);
+        OSSL_CRMF_POPO_free(crm->popo);
+        crm->popo = NULL;
+        return 1;
```

Comment:
Perhaps this case could be a special case `if` before we do the memory allocation above. It seems incorrect to allocate and then immediately deallocate.

Reply:
Good point; done.